instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter.async;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/** A global registry of {@link AsyncOperationEndStrategy} implementations. */
+public final class AsyncEndStrategies {
+  private static final List<AsyncOperationEndStrategy> STRATEGIES = new CopyOnWriteArrayList<>();
```

Comment:
shouldn't this be something weak

Reply:
Actually no, not this one - these strategies shouldn't expire and get removed automatically, once installed they should be always available until unregistered (or the app shuts down)

Reply:
What if the application that contains the library with async strategy gets undeployed? Won't keeping a strong ref here prevent gc of that applications class loader?

Reply:
> What if the application that contains the library with async strategy gets undeployed? Won't keeping a strong ref here prevent gc of that applications class loader?

Oh, that's an excellent point. `AsyncOperationEndStrategy` instances are created just to be registered in this class, they're not being referenced in any other context - a weak-keys CL->strategy map wouldn't work, since the strategy references the CL; a weak-values one wouldn't work either, because strategies could disappear even while they're still needed... Any ideas?

Reply:
The api could say that it is users responsibility to keep strategy alive e.g. strategy instances could be singletons. Or we could build an utility that allows inserting strong reference from class loader to object and use that to bind `AsyncOperationEndStrategy` instance to the class loader that defined it so that we could use weak refs without worrying that the strategy instance gets collected prematurely.

Reply:
Would that work with the ways that `AsyncSpanEndStrategy` is implemented now?  What in the different reactive modules would hold onto the root reference of the strategy?

Reply:
I think `AsyncSpanEndStrategy` has the same problem - the non-enum strategies don't have any singleton instance and putting them in a weak reference would make them disappear.

> What in the different reactive modules would hold onto the root reference of the strategy?

In our javaagent instrumentations we can just create static singleton instances of them and that would solve the problem of weak references, but I have no idea what to do in library instrumentations, like spring-boot-autoconfigure.

Reply:
Yeah, switching the implementation sounds like a reasonable idea. I'll try it out.

Reply:
Done 👍